### PR TITLE
Implement dead letter queue

### PR DESCRIPTION
**Pull request by jnyi**
See https://docs.google.com/document/d/1OiCxRM8-Hq2jvk47aNaLa5zKRbAqBF4KOpiG71_St8M/edit?tab=t.0

Dead letter queue is introduced to avoid retry storms:
1. it adds data that are timed out to main queue
2. it will flush data at each tick
3. if dead letter queue is full, then we drop the samples to avoid retry storms and cascading failures

**What this PR does / why we need it**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note

```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

Unit Test pass:
```
2024-10-14T11:06:19.750-0700	DEBUG	promremote/storage.go:150	Adding a query to the dead letter queue
2024-10-14T11:06:19.751-0700	DEBUG	promremote/storage.go:150	Adding a query to the dead letter queue
2024-10-14T11:06:19.752-0700	DEBUG	promremote/storage.go:150	Adding a query to the dead letter queue
```

### CODE REVIEW

**Review by yuchen-db — APPROVED** [review]

File: src/query/storage/promremote/storage.go [hunk 1]
```diff
@@ -118,6 +118,47 @@ func (wq *WriteQueue) Flush(ctx context.Context, p *promStorage) {
...
+type deadLetterQueue struct {
+	capacity int
+	queries  []*storage.WriteQuery
+
+	sync.Mutex
+}
+
+func NewDeadLetterQueue(capacity int) *deadLetterQueue {
+	return &deadLetterQueue{
+		capacity: capacity,
```
Comment (hczhu-db) [inline src/query/storage/promremote/storage.go:134]: pre-allocate the slice
```
queries: make([]*storage.WriteQuery, 0, capacity)
```

File: src/query/storage/promremote/storage.go [hunk 2]
```diff
@@ -118,6 +118,47 @@ func (wq *WriteQueue) Flush(ctx context.Context, p *promStorage) {
...
+		return errors.New("dead letter queue is full")
+	}
+}
+
+func (dlq *deadLetterQueue) flush(p *promStorage, ctx context.Context, wg *sync.WaitGroup, pendingQuery map[tenantKey]*WriteQueue) {
+	dlq.Lock()
+	defer dlq.Unlock()
+	for _, query := range dlq.queries {
+		p.appendSample(ctx, wg, pendingQuery, query)
+	}
```
Comment (hczhu-db) [inline src/query/storage/promremote/storage.go:165]: Do you need to clear the queue here?
```
dlq.queries = dlq.queries[0:0]
```

Comment (hczhu-db) [reply src/query/storage/promremote/storage.go:165]: I suggest adding some unit test cases for this dead letter queue.

File: src/query/storage/promremote/storage.go [hunk 3]
```diff
@@ -378,13 +420,11 @@ func (p *promStorage) Write(_ context.Context, query *storage.WriteQuery) error
 		p.inFlightSamples.Update(float64(p.inFlightSampleValue.Add(samples)))
 		p.dataQueueSize.Update(float64(len(p.dataQueue)))
 	case <-time.After(*p.opts.queueTimeout):
-		err := errors.New("timeout waiting for data queue for " + p.opts.queueTimeout.String())
-		p.droppedSamples.Inc(samples)
-		if p.isShadowMode {
+		err := p.dlq.add(query)
+		if err != nil {
+			p.droppedSamples.Inc(samples)
 			p.logger.Error("error enqueue samples for prom remote write", zap.Error(err),
```
Comment (hczhu-db) [inline src/query/storage/promremote/storage.go]: Sample this log line. I saw massive log lines in yesterday's incident.

Comment (jnyi) [reply src/query/storage/promremote/storage.go]: ack

**Pull request merged**